cl/citations/tests.py

```diff
@@ -563,6 +575,51 @@ def test_make_html(self):
             )
             print "✓"
 
+    def test_make_html_from_html(self):
+        """Can we convert the HTML of an opinion into modified HTML?"""
+        # fmt: off
+
+        test_pairs = [
+            # Id. citation with HTML tags
+            ('<div><p>the improper views of the Legislature.\" 2 <i>id.,</i> '
+             'at 73.</p>\n<p>Nathaniel Gorham of Massachusetts</p></div>',
+             '<div><p>the improper views of the Legislature." 2<span class="'
+             'citation no-link"> <i><span class="id_token">id.,</span></i> at '
+             '73.</span></p>\n<p>Nathaniel Gorham of Massachusetts</p></div>'),
+
+            # Id. citation with an intervening HTML tag
+            ('<div><p>the improper views of the Legislature.\" 2 <i>id.,</i> '
+             'at <b>73, bolded</b>.</p>\n<p>Nathaniel Gorham of Massachusetts'
+             '</p></div>',
+             '<div><p>the improper views of the Legislature." 2<span class="'
+             'citation no-link"> <i><span class="id_token">id.,</span></i> at '
+             '<b>73, </b></span>bolded.</p>\n<p>Nathaniel Gorham of '
```

Comment:
Hm, I don't think we can change what's bolded in a legal opinion without getting into trouble. That said, I don't know what the preferred HTML would be in this case. Maybe close out the span early?

```
<span class="citation no-link"> <i><span class="id_token">id.,</span></i></span> at <b>73, bolded</b>.
```

Or perhaps we minimize our addition of HTML to the bare minimum, only adding an `<a>` tag? 

Not sure which is the right approach.

Reply:
I agree that "closing the span early" would circumvent this issue (I think), but unfortunately that would require rolling back the changes in #1173, i.e., including the "at XXX" as part of the link. I'm not sure what you mean by only adding an `<a>` tag -- could you explain more?

To be clear, what is currently happening is this:
1. Original HTML:
```
<p>foo <i>id.,</i> at <b>73, asdf ghjk</b>.</p>
```
2. HTML after our regex replacement (note how the injected `span` intersects with the `b` tags):
```
<p>foo<span class="citation no-link"> <i><span class="id_token">id.,</span></i> at <b>73, asdf </span>ghjk</b>.</p>
```
3. After auto clean-up ("moving" the closing `b` tag inside the `span`):
```
<p>foo<span class="citation no-link"> <i><span class="id_token">id.,</span></i> at <b>73, asdf </b></span>ghjk.</p>
```

Note that I haven't actually observed this kind of original HTML in the wild (though I've haven't looked at a ton of opinions); this edge case is just something that occurred to me as a possibility that needed to be dealt with.

Reply:
I think I'm seeing three options:

1. Only do an <a> tag. That is, if we can't linkify it, we don't add HTML at all, because this demonstrates that it's too dangerous to do so — maybe. If we _can_ linkify it, we add the <a> tag and that's it. No more spans. 

2. The way you've got it. Fix it by moving the <b> inside. Bad because it changes semantics. Good because fairly simple.

3. Fix it by moving the span inside if needed. Bad because it doesn't linkify the `at 22` part, but semantics are maintained and HTML validity is preserved.

I think I lean towards option three at this point. I don't think the minimal solution is great, and I think we change the semantics at our peril. I honestly don't understand the regex magic you're doing at this point, and I haven't looked at the HTML recently enough to know how hard it is, but is there a way to do that only when needed? That is, keep the span working as present, but do the fix when there are bold (or other) tags in bad places?

Reply:
I agree that option three might be the best, but that's too complicated for me to implement right now. (I can imagine a way to do it, but it seems formidable.)

So, in b13b67a, I've taken another route. Full linkification/HTML-ification will proceed as normal if the pre-existing HTML is not too complex (read: not unbalanced when we match the Id./Ibid. token). However, if the pre-existing HTMl *is* too complex, we don't replace the HTML for that citation at all.

I think this is a pretty safe/conservative approach, and I actually sort of doubt that there will be many of these edge cases. So even though we're basically just ignoring them, I think they'll be vastly outweighed by the easy cases we do handle. N.B., also, that such citations are still counted by the parser, they just won't be linkified.

If you agree, let me know what you think of my implementation here. I used the `lxml` library because it seemed simple, but maybe there's a better way to parse HTML strings. I also put a utility function in `cl.citations.utils` -- maybe that should go somewhere else?